### PR TITLE
[Filebeat] Add URI Parts Processor to multiple modules

### LINKED ISSUE #19088: Default nginx pipelines should decode url.original field
**Describe the enhancement:**

**Describe a specific use case for the enhancement or feature:**

Here is how `url.original` field currently looks in Kibana:
`/A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4`

Here is how it should look:
`/A Beka G1 Howe/009/17 Reading Elephants.mp4`

Here is the original record in the Nginx access log:
```
lessons.example.com 192.168.0.1 - - [09/Jun/2020:12:10:39 -0700] "GET /A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4 HTTP/1.1" 206 7648063 "http://lessons.example.com/A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4" "Mozilla/5.0 (Linux; Android 5.1.1; KFFOWI) AppleWebKit/537.36 (KHTML, like Gecko) Silk/81.2.16 like Chrome/81.0.4044.138 Safari/537.36"
```
So, that example fixes the issue of HTML-encoded character, such as spaces.

Here is one more example, where decoding is needed to non-english characters:

Here is how `url.original` field currently looks in Kibana:
`/%D0%A0%D1%83%D1%81%D1%81%D0%BA%D0%B0%D1%8F%20%D1%88%D0%BA%D0%BE%D0%BB%D0%B0%20-%20InternetUrok%201%D0%BA%D0%BB%D0%B0%D1%81%D1%81/`

Here is how it should look:
`/Русская школа - InternetUrok 1класс/`

Here is the original record in the Nginx access log:
```
lessons.example.com 192.168.0.1 - - [09/Jun/2020:21:31:51 -0700] "GET /%D0%A0%D1%83%D1%81%D1%81%D0%BA%D0%B0%D1%8F%20%D1%88%D0%BA%D0%BE%D0%BB%D0%B0%20-%20InternetUrok%201%D0%BA%D0%BB%D0%B0%D1%81%D1%81/ HTTP/1.1" 200 894 "http://lessons.example.com/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36"
```

Issue is the same for Nginx access and error pipelines (message field)

Screen Shot 2020-06-09 at 10 08 24 PM

### PR DESCRIPTION
**Pull request by legoguy1000**
## What does this PR do?
Updates Ingest Pipelines for the below modules:

Apache, Nginx, IIS, Traefik, S3Access, Cisco, F5, Fortinet, Google Workspace, Imperva, Microsoft, Netscout, O365, Sophos, Squid, Suricata, Zeek, Zia, Zoom, ZScaler

With the below changes

- Add `uri_parts` processor to parse URIs (includes URL decoding) to add `url.path`, `url.extension`, `url.query`....
- URL Decodes `http.request.referrer` (when applicable) to make them human readable

## Why is it important?

Parses URLs to break up the URL into the different parts and URL decodes them.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

```
cd beats/filebeat
GENERATE=true TESTING_FILEBEAT_MODULES=apache,nginx,iis,traefik mage -v pythonIntegTest
cd beats/x-pack/filebeat
GENERATE=true TESTING_FILEBEAT_MODULES=s3access,cisco,f5,fortinet,google_workspace,imperva,microsoft,netscout,o365,sophos,squid,suricata,zeek,zia,zoom,zscaler mage -v pythonIntegTest
```

## Related issues

- Closes #19088

### CODE REVIEW

**Comment by legoguy1000** [comment]
@andrewstucki This should be ready review and CI tests

**Comment by andrewstucki** [comment]
jenkins run tests

**Comment by andrewstucki** [comment]
@legoguy1000 so it looks to me like the new log entries that were generated for the nginx module are missing `user_agent.device.type` which was added recently to Elasticsearch in https://github.com/elastic/elasticsearch/pull/69322 -- can you make sure your development Elasticsearch instance is up-to-date with the latest snapshot before regeneration? I believe you should be able to grab the latest docker container with:

```
docker pull docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT
```

and then run it with:

```
docker run -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" -e "logger.org.elasticsearch.transport=trace" docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT
```

prior to re-running the regeneration for `nginx`.

**Comment by legoguy1000** [comment]
> @legoguy1000 so it looks to me like the new log entries that were generated for the nginx module are missing `user_agent.device.type` which was added recently to Elasticsearch in https://github.com/elastic/elasticsearch/pull/69322 -- can you make sure your development Elasticsearch instance is up-to-date with the latest snapshot before regeneration? I believe you should be able to grab the latest docker container with:
> 
> ```
> docker pull docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT
> ```
> 
> and then run it with:
> 
> ```
> docker run -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" -e "logger.org.elasticsearch.transport=trace" docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT
> ```
> 
> prior to re-running the regeneration for `nginx`.

Not a problem

File: filebeat/module/iis/error/test/iis_error_url.log-expected.json
Comment (andrewstucki) [inline filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: this seems wrong -- I believe that the backslashes in the original event are doubled due to escaping and this seems to escape the escaped backslashes further?

**Review by andrewstucki — COMMENTED** [review]
Question about the backslash behavior. Not entirely sure if this is behavior based off of a combination of using `urldecode` + `uri_parts` and ordering or if this is just some strange behavior/potentially a bug in the processor itself. I'll try and check this out locally and play around with it to see what's going on.

File: filebeat/module/iis/error/test/iis_error_url.log-expected.json
Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: I'll take a look on my end as well

**Comment by legoguy1000** [comment]
Ran the pipelines to update the `user_agent.device.type` taking a look at the extra escapes.

File: filebeat/module/iis/error/test/iis_error_url.log-expected.json
Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: definitely appears to be the `uri_parts` processor, not the `url_decode`

Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: looks like it may be how it reads the field and is doing something like https://stackoverflow.com/a/23363356/15496837

Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: it looks like any processor thats reading the string and doing anything with it is being double escaped by Java.  even just 
```
- set:
    field: url.original
    value: "{{_tmp.url_orig}}"
    ignore_failure: true
    if: ctx?._tmp?.url_orig != null && ctx?.url?.original == null
```
is causing the `"url.original": "..\\\\..\\\\..\\\\..\\\\..\\\\..\\\\winnt\\\\win.ini"`

Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestDocument.java#L106 and https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/ingest/IngestDocument.java#L648

Comment (andrewkroh) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: The templated value `{{_tmp.url_orig}}` will be escaped. That's why `set` now as a `copy_from` option. You can use it instead of a value template and it does an exact string copy with no escaping added.

Comment (andrewkroh) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: And apparently `{{{foo}}}` (three braces) disables the escaping. https://mustache.github.io/mustache.5.html

That's probably a better way to go in Filebeat since it has better backward compatibility than `copy_from`.

Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json:160]: ok will try that

Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json]: that worked. Updated.

**Comment by legoguy1000** [comment]
@andrewstucki @andrewkroh If you guys are good, can u run the pipeline?

File: filebeat/module/iis/error/test/iis_error_url.log-expected.json
Comment (andrewstucki) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json]: Great, by the way, opened up https://github.com/elastic/elasticsearch/issues/71359 to document the template value interpolation HTML-escaping behavior -- learned something new today 😅

**Comment by andrewstucki** [comment]
jenkins run tests

File: filebeat/module/iis/error/test/iis_error_url.log-expected.json
Comment (legoguy1000) [reply filebeat/module/iis/error/test/iis_error_url.log-expected.json]: Ya definitely something to make sure is documented. :)

File: filebeat/module/apache/access/test/ssl-request.log-expected.json
Comment (andrewstucki) [inline filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: so, I'm actually wondering if we want to run this through the `urldecode` processor or not, it seems to me like we'd want to keep the url encoding, at least for the original field, and, if not, we'll need to document this as a breaking change

File: filebeat/module/iis/access/test/test-x-forward-for.log-expected.json
Comment (andrewstucki) [inline filebeat/module/iis/access/test/test-x-forward-for.log-expected.json:219]: since this is a result of the switch to the `uri_parts` decoder, I'm fine with the url-decoded values, but I think we'll need to document that (potentially apart from `url.original`) the `url.*` fields are now all decoded since it's a breaking change.

**Review by andrewstucki — COMMENTED** [review]
So, based off of https://github.com/elastic/beats/issues/19088#issuecomment-806641864 I want to see if @ebeahan can chime in about whether the best course is to keep the `url.original` field in url-encoded form or if it should be decoded.

File: filebeat/module/iis/access/test/test-x-forward-for.log-expected.json
Comment (legoguy1000) [reply filebeat/module/iis/access/test/test-x-forward-for.log-expected.json:219]: So if we want to say ALL `url.*` fields, we'd have to update the remining pipelines.  I only touched the ones that had sample data and could validate the change.  There were a bunch that had the "auto generated" JS pipelines from the rsa2elk that had actions to parse a url field but no sample data to show so I didn't touch those. I can go though and update those too if we want?

Comment (andrewstucki) [reply filebeat/module/iis/access/test/test-x-forward-for.log-expected.json:219]: Yeah, I'd be fine just documenting it as a single line item with a list of all the modules that were updated under the `filebeat` section in the changelog, but lest the wording needs to change, may want to hold off until there's consensus about `url.original`

File: filebeat/module/apache/access/test/ssl-request.log-expected.json
Comment (legoguy1000) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: So interpreted the comment to say if the URI parts process is decoding all the parts then the `url.original` field should match.  If not, easy kill to remove the decode.  The reason i'm thinking to decode is because the user may enter a unencoded url and the "bad lists" and from what i've seen in the sample data, applications are inconsistent in how they log, some urlencode some have the decoded text. If the intent of the field is to be exactly what the system logged, then i say yes remove the processor.  but if we want it to be consistent across modules to enable correlation, I think you have to decode it.

Comment (andrewstucki) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: yeah, so the intent of the `url.original` field is the core of the question here, based off of the documentation for `url.original` I'm thinking it's supposed be encoded given https://www.elastic.co/guide/en/ecs/current/ecs-url.html#field-url-original says:

> Unmodified original url as seen in the event source.

which I would take to mean that it should maintain the encoding.

Comment (legoguy1000) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: so however the system logs it, just keep it that way? and then you use the other `url.*` fields to do the correlation on since they're decoded

Comment (andrewstucki) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: that's my thought -- that if you need correlation in a standardized way, then you'd use the normalized fields, I'm thinking that `url.full` is likely the normalized version of `url.original`: https://www.elastic.co/guide/en/ecs/current/ecs-url.html#field-url-full

Comment (legoguy1000) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: So I have https://github.com/elastic/elasticsearch/pull/70947 because the URI Java object doesn't like non-url encode spaces or other items in the string and fails to parse, but the URL object handles it.  I guess I should code to generate `url.full` from the decoded parts.  Thoughts?

Comment (andrewstucki) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: Yeah, my thought is that what you probably want here is to maintain the url encoding in `url.original` and potentially extend the processor on the elasticsearch side (or submit a feature request, or both) to add `url.full` as an additional field that gets filled. But before doing that work I'd definitely wait for the ECS core folks to chime in to make sure that that would be the proper field to leverage for full url normalization.

Comment (ebeahan) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: I agree that `url.original` should maintain the encoding. ECS specifies that `url.original` should be the `unmodified original url as seen in the event source`, and I would interpret that to include when the URL is encoded in the original source event.

`url.full` is aimed at sources that may not natively provide the entire, full URL in the original source, but it may be useful to reconstruct it (nginx access logs that separate path, host, protocol, etc.). Since `url.full` doesn't specify unmodified or not, I think adding the decoded URL to `url.full` works.

Comment (legoguy1000) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: Ok so remove the urldecodes. I'll try to knock that out a little later.

Comment (legoguy1000) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json:54]: SO I also added url decode processors for the `http.request.referer` field because it tends to be encoded. The documentation doesn't say anything bout that and the prime example of why was for stuff like `http://lessons.example.com/A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4` -> `http://lessons.example.com/A Beka G1 Howe/029_AND_30/15 reading elephants.mp4`. Should I remove those too or leave them?

Comment (legoguy1000) [reply filebeat/module/apache/access/test/ssl-request.log-expected.json]: Removed all the `url.original` URL decodes.  Add `uri_parts` to a few more modules (Zeek, Zia, Zoom, Google Workspace, Microsoft)

**Comment by andrewstucki** [comment]
jenkins run tests

**Comment by legoguy1000** [comment]
@andrewstucki I fixed the error from the pipeline.  Should be good to test again.

**Comment by andrewstucki** [comment]
jenkins run tests

File: CHANGELOG.next.asciidoc
Comment (andrewstucki) [inline CHANGELOG.next.asciidoc:823]: can you add one additional changelog entry that specifies the url encoding change (due to using the `uri_parts` processor) under the "Breaking Changes" section for "Filebeat"? It can say something to the effect of:

> All `url.*` fields apart from `url.original`  in the ___, ____, and ____ modules are now url unescaped due to using the Elasticsearch  `uri_parts` processor. {pull}24699[24699]

After that I'm good with merging this.

Comment (legoguy1000) [reply CHANGELOG.next.asciidoc:823]: Will do.

Comment (legoguy1000) [reply CHANGELOG.next.asciidoc:823]: I still have the url_decode for the `http.request.referer`. Any thoughts on if I should remove it or keep it?

Comment (legoguy1000) [reply CHANGELOG.next.asciidoc:823]: @andrewstucki I will add the change to the log file today or tomorrow. Any thoughts on the question in the above comment??

Comment (legoguy1000) [reply CHANGELOG.next.asciidoc:839]: Updated the changelog, just want to finalize the `http.request.referer` question and then we should be good to merge.

Comment (andrewstucki) [reply CHANGELOG.next.asciidoc:840]: Sorry for the delayed response, took a look at the ECS docs and they don't appear to specify one way or the other about url decoding. Personally I'd feel like it'd be better to leave them encoded in whatever format they came in, but I could go either way.

**Comment by andrewstucki** [comment]
jenkins run tests

File: CHANGELOG.next.asciidoc
Comment (legoguy1000) [reply CHANGELOG.next.asciidoc:840]: ok, i'll remove it for now so we can get this merged and can always revisit.

Comment (legoguy1000) [reply CHANGELOG.next.asciidoc:840]: @andrewstucki I removed the `url_decode` for `http.request.referrer` field and added the additional entry to the change log.  Should be ready for review.

**Comment by andrewstucki** [comment]
jenkins run tests

**Comment by legoguy1000** [comment]
@andrewstucki I think we are good to merge unless you had any other comments?

**Comment by andrewstucki** [comment]
jenkins run tests

**Comment by andrewstucki** [comment]
@legoguy1000 In general, this looks fine to me, I'll try and get it merged assuming that the tests pass

**Comment by legoguy1000** [comment]
> @legoguy1000 In general, this looks fine to me, I'll try and get it merged assuming that the tests pass

I suspect its going to fail since they made the changes to the geo IP database, i'm re-generating the data now.

**Comment by legoguy1000** [comment]
@andrewstucki I update the generated data files to account for the changes in the pipeline.  If you rerun the tests, we should be good.

**Comment by andrewstucki** [comment]
jenkins run tests

**Comment by legoguy1000** [comment]
@andrewstucki all passed

**Review by andrewstucki — APPROVED** [review]

**Comment by andrewstucki** [comment]
@legoguy1000 thanks for the additions, I'll backport this for the 7.14 release

**Comment by legoguy1000** [comment]
> @legoguy1000 thanks for the additions, I'll backport this for the 7.14 release

Ya, that was definitely a long one.  Definitely good conversations with the ECS team clarifying the fields.

**Comment by andrewstucki** [comment]
@legoguy1000 backport opened at https://github.com/elastic/beats/pull/25353

**Pull request merged**
